staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,612 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queueset
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apiserver/pkg/util/flowcontrol/counter"
+	fq "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing"
+	"k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	"k8s.io/apiserver/pkg/util/shufflesharding"
+	"k8s.io/klog"
+)
+
+// queueSetFactoryImpl implements the QueueSetFactory interface
+// queueSetFactoryImpl makes QueueSet objects.
+type queueSetFactoryImpl struct {
+	// counter can be nil and is ignored in that case
+	counter counter.GoRoutineCounter
+
+	clk clock.PassiveClock
+}
+
+// NewQueueSetFactory creates a new QueueSetFactory object
+func NewQueueSetFactory(clk clock.PassiveClock, counter counter.GoRoutineCounter) fq.QueueSetFactory {
+	return &queueSetFactoryImpl{
+		counter: counter,
+		clk:     clk,
+	}
+}
+
+// NewQueueSet creates a new QueueSet object
+// There is a new QueueSet created for each priority level.
+func (qsf queueSetFactoryImpl) NewQueueSet(name string, concurrencyLimit,
+	desiredNumQueues, queueLengthLimit, handSize int,
+	requestWaitLimit time.Duration) (fq.QueueSet, error) {
+	return newQueueSetImpl(name, concurrencyLimit, desiredNumQueues,
+		queueLengthLimit, handSize, requestWaitLimit, qsf.clk, qsf.counter)
+}
+
+// queueSetImpl is a fair queuing implementation designed for the kube-apiserver.
```

Comment:
Actually it is not specific to the kube-apiserver, not even in our intended usage (all apiservers).  This package is usable for other kinds of servers too.

Reply:
Removed the `kube-apiserver` statement.  Now reads:
```
// queueSetImpl is a fair queuing implementation designed with three major differences:
```